pkg/http/sse_test.go

```diff
@@ -0,0 +1,281 @@
+package http_test
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	netHTTP "net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill-http/pkg/http"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/pubsub/gochannel"
+	"github.com/go-chi/chi"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSSE(t *testing.T) {
+	router, err := message.NewRouter(message.RouterConfig{}, watermill.NopLogger{})
+	require.NoError(t, err)
+
+	pubsub := gochannel.NewGoChannel(gochannel.Config{}, watermill.NopLogger{})
+
+	sseRouter, err := http.NewSSERouter(router, pubsub, http.DefaultErrorHandler, watermill.NopLogger{})
+	require.NoError(t, err)
+
+	postUpdatedTopic := "post-updated"
+
+	allPostsHandler := sseRouter.AddHandler(postUpdatedTopic, allPostsStreamAdapter{
+		allPostsRepository: postsRepositoryMock{},
+	})
+
+	postHandler := sseRouter.AddHandler(postUpdatedTopic, postStreamAdapter{
+		postsRepository: postsRepositoryMock{},
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		err := router.Run(ctx)
+		if err != nil {
+			t.Errorf("unexpected error while startig router: %v", err)
+		}
+	}()
+
+	r := chi.NewRouter()
+	r.Get("/posts", allPostsHandler)
+	r.Get("/posts/{id}", postHandler)
+
+	server := httptest.NewServer(r)
+	defer server.Close()
+
+	postID := watermill.NewUUID()
+
+	t.Run("generic_request", func(t *testing.T) {
+		req, err := netHTTP.NewRequest("GET", server.URL+"/posts/"+postID, nil)
+		require.NoError(t, err)
+
+		resp, err := netHTTP.DefaultClient.Do(req)
+		require.NoError(t, err)
+
+		require.Equal(t, 200, resp.StatusCode)
+		body, err := ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		post := Post{}
+		err = json.Unmarshal(body, &post)
+		require.NoError(t, err)
+		require.Equal(t, postID, post.ID)
+	})
+
+	t.Run("event_stream_no_updates", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+
+		responses := newSSERequest(t, ctx, server.URL+"/posts/"+postID)
+
+		response := <-responses
+		post := Post{}
+		err = json.Unmarshal(response, &post)
+		require.NoError(t, err)
+		require.Equal(t, postID, post.ID)
+
+		response = <-responses
```

Comment:
I can't see what is writing to this channel? won't those channel reads freeze sometimes?

Reply:
See https://github.com/ThreeDotsLabs/watermill-http/pull/3/files/e2b23092f2a8a94337b7ac4bb3d50e307ffbe611#diff-2da0048a18646a03babdb8150c78f27fR187